### ISSUE TITLE
"Reduce UI motion" option

### ISSUE THREAD

**Issue by ilianaw** [tootsuite/mastodon#5355]
Hello from witches.town, where the version is currently 2.0.0rc2.

The post privacy selector under the toot authoring box now has a noticeable animation, which may be from either #5018 or #5140. This has made me experience motion sickness a couple of times now. :(

I'd like to request an option to reduce/remove UI motion. This is similar to disabling GIF animation by default, but I think it's still useful to have them as separate options (I think most folks who want to disable GIF animation for non-accessibility reasons may want to keep UI motion on).

I use the term "reduce motion" because that's what my phone uses in the accessibility menu. Something else might make more sense. In the sense of iOS's reduce motion feature, it replaces most of the scaling animations with simple crossfades.

There are some other parts of UI motion that haven't bothered me yet but may be bothering others:
* Boost/fav button animations (could fade in instead of animate)
* Progress bar that displays at the top while waiting for things to load / operations to complete

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [x] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by nolanlawson** [comment]
Out of curiosity, are you using a browser that supports [prefers-reduced-motion](https://webkit.org/blog/7551/responsive-design-for-motion/), e.g. Safari? If so we could potentially hook into that.

Although I kinda suspect that given it's only a CSS draft spec and only available in Safari, for the time being we should probably do the user option instead.

**Comment by nightpool** [comment]
yes we should support the user option—the browser option is good but it's too far out for us to hold it up as a viable solution in good faith.

**Comment by nightpool** [comment]
I'm interested in your opinion on the CW animation: https://imgur.com/a/VpGr5

it seems to be in the similar class of motion that should be reduced, but *something* needs to move to create space for the CW box to appear in, and intuitively I don't think that making it just an instant jerk would help the situation much. A fade isn't possible here since there's only one thing "appearing", but the content box has to move for there to be *space* for that thing to appear.

thoughts?

**Comment by nolanlawson** [comment]
Most of the fancy animations are [ReactMotion](https://github.com/chenglou/react-motion). It doesn't seem to have a built-in option to disable animations, so maybe we could wrap all access to it to check a user option first.

As for CSS animations, those would be much harder to disable/enable from JavaScript, but it sounds like those aren't the ones causing motion sickness (perhaps because they're simpler?).

**Comment by nightpool** [comment]
@nolanlawson adding a "has-motion" body class in ruby and gating our CSS animations on that seems pretty easy.

**Comment by nolanlawson** [comment]
Hmmm for people with vestibular spectrum disorder, though, is it valuable to disable _all_ animations? E.g. I wonder for the standard `ease-in-out` CSS opacity transforms on the media modal, if we made it "pop in" instantly instead, would that be worse or better?

I really don't know much about this disorder aside from what I've read in the WebKit article above, so might be good to do some research first. But at the very least I think we can limit the fancy ReactMotion animations which have already been confirmed as nausea-inducing.

**Comment by ilianaw** [comment]
> Out of curiosity, are you using a browser that supports prefers-reduced-motion, e.g. Safari? If so we could potentially hook into that.

Only on mobile. [Firefox's request for it](https://bugzilla.mozilla.org/show_bug.cgi?id=1365045) appears to not be prioritized currently.

As I commented in #5393 I think it's useful to support both, but maybe not right out of the gate.

> I'm interested in peoples opinions on the CW animation: https://imgur.com/a/VpGr5

To me the CW animation is fine but I'm also not everybody... I'm not sure what else you could do.

**Comment by nolanlawson** [comment]
Fixed by #5355, please tell me if I missed any nauseating motions, but now that it's wrapping `react-motion` everywhere I hopefully got them all!
